README.md
[![NPM](https://nodei.co/npm/yocto-sitemap-generator.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/yocto-sitemap-generator/)

![alt text](https://david-dm.org/yoctore/yocto-sitemap-generator.svg "Dependencies Status")
[![Code Climate](https://codeclimate.com/github/yoctore/yocto-sitemap-generator/badges/gpa.svg)](https://codeclimate.com/github/yoctore/yocto-sitemap-generator)
[![Test Coverage](https://codeclimate.com/github/yoctore/yocto-sitemap-generator/badges/coverage.svg)](https://codeclimate.com/github/yoctore/yocto-sitemap-generator/coverage)
[![Issue Count](https://codeclimate.com/github/yoctore/yocto-sitemap-generator/badges/issue_count.svg)](https://codeclimate.com/github/yoctore/yocto-sitemap-generator)
[![Build Status](https://travis-ci.org/yoctore/yocto-sitemap-generator.svg?branch=master)](https://travis-ci.org/yoctore/yocto-sitemap-generator)

## Overview

This module is a part of yocto node modules for NodeJS.

Please see [our NPM repository](https://www.npmjs.com/~yocto) for complete list of available tools (completed day after day).

This module provide a simple sitemap generator based [sitemap-generator](https://www.npmjs.com/package/sitemap-generator).

## Motivation

Create an simple module that generate automatically sitemap by crawling an site.

## Specification

The generation can be done directly or handle by an scheduler.

- directly : call process() method to generate the sitemap
- scheduler : scheduler will generate sitemap automatically. The module [node-scheduler](https://www.npmjs.com/package/node-schedule) was used.


## Usage

### Basic configuration

> Configuration object will be check by an Joi schema

```javascript

var config = {
  // the path file to write the xml file, optional (default : stiemap.xml)
  pathFile : 'sitemap.xml',
  // Optional, if is not defined or empty the scheduler will not be set
  scheduler  : {
    // Use only cron syntax
    cronRule : '* * * * * *'
  },
  generator : {
    // the url to map
    url : 'http://yocto.re',
    // Optional options
    options : {
      restrictToBasepath: false,
      stripQuerystring: true,
    }
  },
  // Rules to apply to each url found
  rules : {
    // data use into urlset attribute, all is optional (default value below)
    // if value was null, the attribute will not be added into xml file
    urlset : {
      xmlns           : 'http://www.sitemaps.org/schemas/sitemap/0.9',
      xsi             : 'http://www.w3.org/2001/XMLSchema-instance',
      schemaLocation  : 'http://www.sitemaps.org/schemas/sitemap/0.9  http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd'
    },
    // default config applied
    default : {
      params : {
        changefreq : 'weekly',
        priority   : 0.85
      }
    },
    // config to override default params when an route match the regexp
    override : [
      {
        // String that will be use to create a new RegExp
        regex   : 'nos-travaux',
        params  : {
          changefreq : 'daily',
          priority   : 0.5
        }
      },
      {
        regex   : 'nos-services',
        params  : {
          changefreq : 'daily',
          priority   : 1
        }
      }
    ],
    // all route that will be exclude in sitemap
    exclude : [
      // String that will be use to create a new RegExp
      'contact'
    ]
  }
}

var sitemapGenerator = require('yocto-sitemap-generator');

// load config into module
sitemapGenerator.load(config).then(function (value) {
  // loading success
  console.log(value);
}).catch(function (error) {
  // loading failed
  console.log(error);
});

```

### Generation

> First load module with load() method

```javascript

// generate sitemap with config defined
// The first param indicate if the sitemap should be converted to xml file
sitemapGenerator.process(true).then(function (sitemap) {
  // ge success sitemap to xml
  console.log(value);
}).catch(function (sitemap) {
  // loading failed
  console.log(error);
});
```

### Scheduler

> First load module with load() method
"

```javascript

// Enable scheduler that will generate automatically the sitemap
sitemapGenerator.enableScheduler().then(function (sitemap) {
  // enable succes
  console.log(sitemap);
}).catch(function (error) {
  // enable failed
  console.log(error);
});
```


CHANGELOG.md
## 0.1.0 (2016-12-06)
  - Init module


test/schema.js
var generator = require('../src/')();
var assert = require('assert');
var _      = require('lodash');
var utils  = require('yocto-utils');
var should = require('chai').should();

generator.logger.disableConsole();

var config = {
  scheduler  : {
    cronRule : '* * * * * * *'
  },
  generator : {
    url : 'http://yocto.re'
  },
  rules : {
    urlset : {
      xmlns           : 'http://www.sitemaps.org/schemas/sitemap/0.9',
      xsi             : 'http://www.w3.org/2001/XMLSchema-instance',
      schemaLocation  : 'http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd'
    },
    default : {
      params : {
        changefreq : 'weekly',
        priority   : 0.85
      }
    },
    override : [
      {
        regex   : 'nos-travaux',
        params  : {
          changefreq : 'daily',
          priority   : 0.5
        }
      },
      {
        regex   : 'nos-services',
        params  : {
          changefreq : 'daily',
          priority   : 1
        }
      }
    ],
    exclude : [
      'contact'
    ]
  }
};

describe('[ yocto-sitemap-generator ] - test with predefined value', function() {

  // test an valid config
  describe('Load() use correct data for schema ', function() {
    it('An success promise should be return', function (done) {
      generator.load(config, 'load').then(function (value) {
        assert(true);
        done();
      }).catch(function (error) {
        // reject error
        done(error);
      });
    });
  });
});


test/factory.js
var factory = require('../src/modules/factory/index.js')();
var assert = require('assert');
var _      = require('lodash');
var utils  = require('yocto-utils');
var should = require('chai').should();

factory.logger.disableConsole();

var sitemap = {
  urlset: {
    xmlns: 'http://www.sitemaps.org/schemas/sitemap/0.9',
    url: [{
      loc: 'http://www.yocto.re/'
    }, {
      loc: 'http://www.yocto.re/nos-services'
    }, {
      loc: 'http://www.yocto.re/nos-travaux'
    }, {
      loc: 'http://www.yocto.re/contact'
    }, {
      loc: 'http://www.yocto.re/nos-travaux/bazar-maurice-mu'
    }, {
      loc: 'http://www.yocto.re/nos-travaux/bfcoi'
    }, {
      loc: 'http://www.yocto.re/nos-travaux/chambre-de-metiers-et-de-l-artisanat-de-la-reunion'
    }]
  }
};

var baseUrl = 'http://wwww.yocto.re';

var config = {
  urlset : {
    xmlns           : 'http://www.sitemaps.org/schemas/sitemap/0.9',
    xsi             : 'http://www.w3.org/2001/XMLSchema-instance',
    schemaLocation  : 'http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd'
  },
  default : {
    params : {
      changefreq : 'weekly',
      priority   : 0.85
    }
  },
  override : [
    {
      regex   : 'nos-travaux',
      params  : {
        changefreq : 'daily',
        priority   : 0.5
      }
    },
    {
      regex   : 'nos-services',
      params  : {
        changefreq : 'daily',
        priority   : 0.8
      }
    }
  ],
  exclude : [
    'contact'
  ]
};

var urls = [
  {
    testCase : 'default',
    initial : {
      loc: 'http://www.yocto.re/nos-projet'
    },
    final : {
      loc        : 'http://www.yocto.re/nos-projet',
      changefreq : 'weekly',
      priority   : 0.85
    }
  },
  {
    testCase : 'orverride',
    initial : {
      loc: 'http://www.yocto.re/nos-travaux'
    },
    final : {
      loc        : 'http://www.yocto.re/nos-travaux',
      changefreq : 'daily',
      priority   : 0.5
    }
  },
  {
    testCase : 'exclude',
    initial : {
      loc: 'http://www.yocto.re/contact'
    },
    final : null
  }
];

var sitemapFinal = "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<urlset xmlns=\"http://www.sitemaps.org/schemas/sitemap/0.9\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd\">\n  <url>\n    <loc>http://www.yocto.re/</loc>\n    <changefreq>weekly</changefreq>\n    <priority>0.85</priority>\n  </url>\n  <url>\n    <loc>http://www.yocto.re/nos-services</loc>\n    <changefreq>daily</changefreq>\n    <priority>0.8</priority>\n  </url>\n  <url>\n    <loc>http://www.yocto.re/nos-travaux</loc>\n    <changefreq>daily</changefreq>\n    <priority>0.5</priority>\n  </url>\n  <url>\n    <loc>http://www.yocto.re/nos-travaux/bazar-maurice-mu</loc>\n    <changefreq>daily</changefreq>\n    <priority>0.5</priority>\n  </url>\n  <url>\n    <loc>http://www.yocto.re/nos-travaux/bfcoi</loc>\n    <changefreq>daily</changefreq>\n    <priority>0.5</priority>\n  </url>\n  <url>\n    <loc>http://www.yocto.re/nos-travaux/chambre-de-metiers-et-de-l-artisanat-de-la-reunion</loc>\n    <changefreq>daily</changefreq>\n    <priority>0.5</priority>\n  </url>\n</urlset>";

describe('[ Factory ] - test with predifined result', function() {
  describe('remapSitemap() - Create sitemap with all config', function() {
    it('Value should be exactly the same value as finalSitemap', function (done) {
      factory.remapSitemap(sitemap, config, baseUrl, true).then(function (value) {
        assert.equal(value, sitemapFinal);
        done();
      }).catch(function (error) {
        // reject error
        done(error);
      });
    });
  });

  // test each config
  _.each(urls, function (v) {
    describe('buildUrl() - rebuild url object with config file - testCase : #' + v.testCase , function () {
      it('Value should be exactly the same value as final ', function (done) {
        factory.buildUrl(v.initial, config, baseUrl).then(function (value) {
          assert.deepEqual(value, v.final);
          done();
        }).catch(function (error) {
          // reject error
          done(error);
        });
      });
    });
  });

});


test/generator.js
var schema = require('../src/modules/schema/')();
var assert = require('assert');
var _      = require('lodash');
var utils  = require('yocto-utils');
var should = require('chai').should();

schema.logger.disableConsole();

var config = {
  scheduler  : {
    cronRule : '* * * * * * *'
  },
  generator : {
    url : 'http://yocto.re'
  },
  rules : {
    urlset : {
      xmlns           : 'http://www.sitemaps.org/schemas/sitemap/0.9',
      xsi             : 'http://www.w3.org/2001/XMLSchema-instance',
      schemaLocation  : 'http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd'
    },
    default : {
      params : {
        changefreq : 'weekly',
        priority   : 0.85
      }
    },
    override : [
      {
        regex   : 'nos-travaux',
        params  : {
          changefreq : 'daily',
          priority   : 0.5
        }
      },
      {
        regex   : 'nos-services',
        params  : {
          changefreq : 'daily',
          priority   : 1
        }
      }
    ],
    exclude : [
      'contact'
    ]
  }
};

describe('[ Schema ] - test with predefined value', function() {

  // test an valid config
  describe('validate() use correct data for schema < load >', function() {
    it('An success promise should be return', function (done) {
      schema.validate(config, 'load').then(function (value) {
        assert(true);
        done();
      }).catch(function (error) {
        // reject error
        done(error);
      });
    });
  });

  // test an valid config
  describe('validate() use wrong data for schema < load >', function() {
    it('An failed promise should be return', function (done) {
      schema.validate({}, 'load').then(function (value) {
        assert.fail(value);
        done(value);
      }).catch(function (error) {
        // reject error
        assert(true);
        done();
      });
    });
  });



});


example/generate.js
var logger    = require('yocto-logger');
var generator = require('../src/')(logger);
var utils = require('yocto-utils');

var xmlParser = require('xml2json');

var config = {
  generator : {
    url : 'http://yocto.re'
  },
  rules : {
    default : {
      params : {
        changeFreq : 'weekly',
        priority   : 0.85
      }
    },
    override : [
      {
        regex   : 'nos-travaux',
        params  : {
          changeFreq : 'daily',
          priority   : 0.5
        }
      },
      {
        regex   : 'nos-services',
        params  : {
          changeFreq : 'daily',
          priority   : 1
        }
      }
    ],
    exclude : [
      'contact'
    ]
  }
};

// load config
generator.load(config).then(function () {

  console.log('\n --> config success ... ');

  // load config
  generator.process().then(function (value) {

    console.log('\n --> Exemple --> generate success ...  \n ' + utils.obj.inspect(value));
    console.log('\n --> Exemple -->typeof ' + typeof value);

  }).catch(function (error) {

    console.log('\n --> generate -  error : ', error);
  });

}).catch(function (error) {

  console.log('\n --> error : ', error);
});


example/scheduler.js
var logger    = require('yocto-logger');
var generator = require('../src/')(logger);
var utils = require('yocto-utils');

var xmlParser = require('xml2json');

var config = {
  scheduler  : {
    cronRule : '*/50 * * * * *'
  },
  generator : {
    url : 'http://yocto.re'
  },
  rules : {
    default : {
      params : {
        changefreq : 'weekly',
        priority   : 0.85
      }
    },
    override : [
      {
        regex   : 'nos-travaux',
        params  : {
          changefreq : 'daily',
          priority   : 0.5
        }
      },
      {
        regex   : 'nos-services',
        params  : {
          changefreq : 'daily',
          priority   : 1
        }
      }
    ],
    exclude : [
      'contact'
    ]
  }
};

// load config
generator.load(config).then(function () {

  console.log('\n --> config success ... ');

  // load config
  generator.enableScheduler().then(function (value) {

    console.log('\n --> Exemple --> generate success ...  \n ' + utils.obj.inspect(value));
    console.log('\n --> Exemple -->typeof ' + typeof value);

  }).catch(function (error) {

    console.log('\n --> generate -  error : ', error);
  });
}).catch(function (error) {

  console.log('\n --> error : ', error);
});


src/modules/schema/index.js
'use strict';

var logger = require('yocto-logger');
var _      = require('lodash');
var utils  = require('yocto-utils');
var Q      = require('q');
var joi    = require('joi');

/**
 *
 * Utility tool to automaticly generate sitemap
 *
 * @date : 06/12/2016
 * @author : Cédric BALARD <cedric@yocto.re>
 * @copyright : Yocto SAS, All right reserved
 *
 * @module Schema
 */
function Schema (l) {

  /**
  * Default logger instance
  *
  * @property logger
  * @type Object
  */
  this.logger = l;
}

/**
 * Method that use joi to validate an object with it schema
 *
 * @param  {Object} data the data to check
 * @param  {Object} schemaName the name of the schema to use
 * @return {Object} promise of this method
 */
Schema.prototype.validate = function (data, schemaName) {
  // create async process
  var deferred  = Q.defer();

  this.logger.debug('[ Schema.validate ] - validate schema for schemaName = ' + schemaName);
  // make joi validation
  var result = joi.validate(data, this.getSchema(schemaName));

  // check if joi validation failed
  if (_.isEmpty(result.error)) {

    // success
    deferred.resolve(result.value);
  } else {

    // Log error
    this.logger.error('[ Schema.validate ] - joi validation failed, details : ' +
    utils.obj.inspect(result.error));

    // reject error
    deferred.reject(result.error);
  }

  // return result of control flow
  return deferred.promise;
};

/**
 * Method that retrieve an schema
 *
 * @param  {String} schemaName the name of schema
 * @return {Object} return the joi object
 */
Schema.prototype.getSchema = function (schemaName) {

  var paramRules = joi.object().optional().keys({
    changefreq  : joi.string().empty().allow([ 'always', 'hourly', 'daily', 'weekly',
    'monthly', 'yearly', 'never' ]),
    priority    : joi.number().optional().min(0).max(1),
    lastmod     : joi.string().optional().trim().empty()
  }).default({});

  // Default urlset value
  var defaultUrlset = {
    xmlns           : 'http://www.sitemaps.org/schemas/sitemap/0.9',
    xsi             : 'http://www.w3.org/2001/XMLSchema-instance',
    schemaLocation  : 'http://www.sitemaps.org/schemas/sitemap/0.9 ' +
    'http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd'
  };

  // All schemas
  var schemas = {
    load            : joi.object().required().keys({
      pathFile  : joi.string().optional().trim().empty().default('sitemap.xml'),
      scheduler : joi.object().optional().keys({
        cronRule : joi.string().required().trim().empty(),
      }).default({}),
      generator : joi.object().required().keys({
        url     : joi.string().trim().required().empty(),
        options : joi.object().optional().keys({
          restrictToBasepath : joi.boolean().optional(),
          stripQuerystring   : joi.boolean().optional()
        }).default({})
      }),
      rules     : joi.object().optional().keys({
        urlset    : joi.object().optional().keys({
          xmlns           : joi.string().trim().empty().default(defaultUrlset.xmlns).allow(null),
          xsi             : joi.string().trim().empty().default(defaultUrlset.xsi).allow(null),
          schemaLocation  : joi.string().trim().empty().default(defaultUrlset.schemaLocation)
          .allow(null),
        }).default(defaultUrlset),
        default   : joi.object().optional().keys({
          params : paramRules
        }).default({
          params : {}
        }),
        override  : joi.array().optional().items(
          joi.object().keys({
            regex  : joi.string().required().trim().empty(),
            params : paramRules
          })
        ).default([]),
        exclude   : joi.array().optional().items(
          joi.string().trim().empty()
        ).default([])
      })
    })
  };

  // return this schema
  return schemas[schemaName];
};

// Default export
module.exports = function (l) {
  // is a valid logger ?
  if (_.isUndefined(l) || _.isNull(l)) {
    logger.warning('[ Schema.constructor ] - Invalid logger given. Use internal logger');
    // assign
    l = logger;
  }
  // default statement
  return new (Schema)(l);
};


example/factory.js
var factory = require('../src/modules/factory')();
var utils = require('yocto-utils');

var sitemap = {
  urlset: {
    xmlns: 'http://www.sitemaps.org/schemas/sitemap/0.9',
    url: [{
      loc: 'http://www.yocto.re/'
    }, {
      loc: 'http://www.yocto.re/nos-services'
    }, {
      loc: 'http://www.yocto.re/nos-travaux'
    }, {
      loc: 'http://www.yocto.re/contact'
    }, {
      loc: 'http://www.yocto.re/nos-travaux/bazar-maurice-mu'
    }, {
      loc: 'http://www.yocto.re/nos-travaux/bfcoi'
    }, {
      loc: 'http://www.yocto.re/nos-travaux/chambre-de-metiers-et-de-l-artisanat-de-la-reunion'
    }]
  }
};

var baseUrl = 'http://wwww.yocto.re';

var config = {
  urlset : {
    xmlns           : 'http://www.sitemaps.org/schemas/sitemap/0.9',
    xsi             : 'http://www.w3.org/2001/XMLSchema-instance',
    schemaLocation  : 'http://www.sitemaps.org/schemas/sitemap/0.9 http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd'
  },
  default : {
    params : {
      changefreq : 'weekly',
      priority   : 0.85
    }
  },
  override : [
    {
      regex   : 'nos-travaux',
      params  : {
        changefreq : 'daily',
        priority   : 0.5
      }
    },
    {
      regex   : 'nos-services',
      params  : {
        changefreq : 'daily',
        priority   : 0.8
      }
    }
  ],
  exclude : [
    'contact'
  ]
};

factory.remapSitemap(sitemap, config, baseUrl, true).then(function (value) {

  console.log('\n --> value : \n\n', value);

}).catch(function (error) {

  console.log('\n --> error : ', error)
});
